### PR TITLE
1 allow and limit rotations of items

### LINKED ISSUE #1: Allow and limit rotations of items
### Is your feature request related to a problem? Please describe
In practice most items can be rotated in one of 6 orthogonal rotations in order to allow for denser packing. Sometimes item constraints mean that only a subset of these 6 rotations are allowed. Adding this feature to the environment will increase the size of the action space and make it more challenging for an agent to optimise packing but (with the right generator) allow denser packings to be achieved.

### Describe the solution you'd like
Allow up to 6 rotations of a cuboid item to be allowed. When an item has constraints that allow only a subset of these rotations to be considered valid, we would mask out these invalid actions.

### Describe alternatives you've considered
No alternatives have been considered.

### Additional context
A feature present in many industrial solutions.

---
### Misc
- [x] Check for duplicate requests.

### PR DESCRIPTION
**Pull request by chouakifares**
- Changed shape of arrays items_mask and items_placed from (max_num_items) to (6, max_num_items).
- Added an attribute to the state called nb_items.
- Item mask now changes dynamically during the episode to mask all the items that have been placed.
- Action_mask is now of size (6, max_num_ems, max_num_items).
- Added a ConstrainedRandomGenerator and a ConstrainedToyGenerator.

### CODE REVIEW

File: configs/config.yaml [hunk 1]
```diff
@@ -0,0 +1,12 @@
+defaults:
+    - _self_
+    - env: upgraded_bin_pack  # [bin_pack, cleaner, connector, cvrp, game_2048, graph_coloring, job_shop, knapsack, maze, minesweeper, mmst, multi_cvrp, robot_warehouse, rubiks_cube, snake, sudoku, tetris, tsp]
```
Comment (rsjeffers) [inline configs/config.yaml]: We can discuss naming more thoroughly for when we have all the features together but it might be good to be consistent with the name of the class you've created (ConstrainedBinPack)?

File: jumanji/environments/packing/bin_pack/conftest.py [hunk 2]
```diff
@@ -76,6 +78,106 @@ def __call__(self, key: chex.PRNGKey) -> State:
...
+        container = make_container(TWENTY_FOOT_DIMS)
+        return State(
+            container=container,
+            ems=jax.tree_util.tree_map(
+                lambda x: jnp.array([x] + (self.max_num_ems - 1) * [0], jnp.int32),
+                container,
+            ),
+            ems_mask=jnp.array([True] + (self.max_num_ems - 1) * [False], bool),
+            items=Item(
+                # The 1st and 2nd items have the same shape.
```
Comment (rsjeffers) [inline jumanji/environments/packing/bin_pack/conftest.py:182]: Is this still the case? It looks like the y lengths are different? Why are they matrices? Should they not be 1D arrays,

File: jumanji/environments/packing/bin_pack/conftest.py [hunk 3]
```diff
@@ -95,6 +202,17 @@ def random_generator() -> RandomGenerator:
     return RandomGenerator(max_num_items=20, max_num_ems=80)
 
 
+@pytest.fixture
+def constrained_toy_generator() -> ToyGenerator:
```
Comment (rsjeffers) [inline jumanji/environments/packing/bin_pack/conftest.py]: Should the return type be updated here?

File: jumanji/environments/packing/bin_pack/conftest.py [hunk 4]
```diff
@@ -95,6 +202,17 @@ def random_generator() -> RandomGenerator:
...
+    return ConstrainedToyGenerator()
+
+
+@pytest.fixture
+def constrained_random_generator() -> RandomGenerator:
```
Comment (rsjeffers) [inline jumanji/environments/packing/bin_pack/conftest.py]: Update return type?

File: jumanji/environments/packing/bin_pack/env.py [hunk 5]
```diff
@@ -37,11 +41,16 @@
...
-from jumanji.environments.packing.bin_pack.viewer import BinPackViewer
+from jumanji.environments.packing.bin_pack.viewer import (
+    BinPackViewer,
+    ConstrainedBinPackViewer,
+)
 from jumanji.tree_utils import tree_add_element, tree_slice
 from jumanji.types import TimeStep, restart, termination, transition
 from jumanji.viewer import Viewer
 
+matplotlib.use("TkAgg")
```
Comment (rsjeffers) [inline jumanji/environments/packing/bin_pack/env.py]: what is this being used for?

File: jumanji/environments/packing/bin_pack/env.py [hunk 6]
```diff
@@ -61,16 +70,16 @@ class BinPack(Environment[State]):
         - ems_mask: jax array (bool) of shape (obs_num_ems,)
             indicates the EMSs that are valid.
         - items: `Item` tree of jax arrays (float if `normalize_dimensions` else int32) each of
-            shape (max_num_items,),
+            shape (max_num_items, 6),
```
Comment (rsjeffers) [inline jumanji/environments/packing/bin_pack/env.py:76]: Weren't these going to be flat?

File: jumanji/environments/packing/bin_pack/env.py [hunk 7]
```diff
@@ -535,7 +543,12 @@ def is_action_allowed(
         )(obs_ems, obs_ems_mask, items, items_mask, items_placed)
         return action_mask
 
-    def _pack_item(self, state: State, ems_id: int, item_id: chex.Numeric) -> State:
+    def _pack_item(
```
Comment (rsjeffers) [inline jumanji/environments/packing/bin_pack/env.py]: Was this reformatted automatically by your IDE? Might be worth putting it back to what it was if possible (It doesn't look like anything has functionally changed)

File: configs/config.yaml (hunk 1 above)
Comment (chouakifares) [reply configs/config.yaml]: This is a mistake, I changed the name of my class yesterday before creating the PR and used the IDE's ability to refactor to make sure it changed everywhere. I guess this hasn't changed since it's a string , I'll change this immediately.

File: jumanji/environments/packing/bin_pack/conftest.py (hunk 2 above)
Comment (chouakifares) [reply jumanji/environments/packing/bin_pack/conftest.py:182]: Yes this is still the case, both items have the dimensions (1000, 700, 900). It's in 2D because I didn't want to modify the whole logic and I opted for flattening the observations only when I needed them to be flat .i.e only inside the actor_critic.py file.

File: jumanji/environments/packing/bin_pack/reward.py [hunk 8]
```diff
@@ -53,8 +53,15 @@ def __call__(
         is_done: bool,
     ) -> float:
         del next_state, is_done
-        _, item_id = action
-        chosen_item_volume = item_volume(tree_slice(state.items, item_id))
+        try:
```
Comment (rsjeffers) [inline jumanji/environments/packing/bin_pack/reward.py]: It would be best to check and obtained the correct slice without doing a try catch

File: jumanji/environments/packing/bin_pack/conftest.py (hunk 3 above)
Comment (chouakifares) [reply jumanji/environments/packing/bin_pack/conftest.py]: I'll update it

File: jumanji/environments/packing/bin_pack/types.py [hunk 9]
```diff
@@ -123,13 +164,14 @@ class State:
     container: Container  # leaves of shape ()
     ems: EMS  # leaves of shape (max_num_ems,)
     ems_mask: chex.Array  # (max_num_ems,)
-    items: Item  # leaves of shape (max_num_items,)
-    items_mask: chex.Array  # (max_num_items,)
-    items_placed: chex.Array  # (max_num_items,)
+    items: Item  # leaves of shape (6,max_num_items,)
+    items_mask: chex.Array  # (6,max_num_items)
+    items_placed: chex.Array  # (6,max_num_items,6)
```
Comment (rsjeffers) [inline jumanji/environments/packing/bin_pack/types.py]: Could you add a comment to explain the shape?

File: jumanji/environments/packing/bin_pack/conftest.py (hunk 4 above)
Comment (chouakifares) [reply jumanji/environments/packing/bin_pack/conftest.py]: Done ✅

File: jumanji/environments/packing/bin_pack/env.py (hunk 5 above)
Comment (chouakifares) [reply jumanji/environments/packing/bin_pack/env.py]: For visualization of the packing.

File: jumanji/environments/packing/bin_pack/env.py (hunk 6 above)
Comment (chouakifares) [reply jumanji/environments/packing/bin_pack/env.py:76]: No, I only flatten them inside the actor_critic.py file.

File: jumanji/environments/packing/bin_pack/env.py (hunk 7 above)
Comment (chouakifares) [reply jumanji/environments/packing/bin_pack/env.py]: Done ✅

File: jumanji/environments/packing/bin_pack/types.py [hunk 10]
```diff
@@ -41,6 +41,47 @@ class Item(NamedTuple):
     z_len: chex.Numeric
 
 
+def rotated_items_from_space(space: Space) -> jnp.ndarray:
+    return Item(
+        x_len=jnp.asarray(
+            [
+                # x along X, y along Y, z along Z
+                space.x2 - space.x1,
```
Comment (rsjeffers) [inline jumanji/environments/packing/bin_pack/types.py:49]: Reorder these so that rotations 0,1,2,3,4,5 correspond to the same orientations as those rotations in deeppack

File: jumanji/training/networks/bin_pack/actor_critic.py [hunk 11]
```diff
@@ -80,20 +80,53 @@ def __init__(
         self.model_size = transformer_num_heads * transformer_key_size
 
     def __call__(self, observation: Observation) -> Tuple[chex.Array, chex.Array]:
+        if len(observation.items_mask.shape) == 3:
```
Comment (rsjeffers) [inline jumanji/training/networks/bin_pack/actor_critic.py]: Add an explanatory comment and maybe even do this in its own function

File: jumanji/training/networks/bin_pack/actor_critic.py [hunk 12]
```diff
@@ -80,20 +80,53 @@ def __init__(
...
+            reshaped_items_mask = observation.items_mask.reshape(
+                observation.items_mask.shape[0],
+                observation.items_mask.shape[1] * observation.items_mask.shape[2],
```
Comment (rsjeffers) [inline jumanji/training/networks/bin_pack/actor_critic.py]: would -1 work here? (and below) for reshaping.

File: jumanji/training/networks/bin_pack/actor_critic.py [hunk 13]
```diff
@@ -115,7 +148,6 @@ def __call__(self, observation: Observation) -> Tuple[chex.Array, chex.Array]:
                 model_size=self.model_size,
                 name=f"self_attention_items_block_{block_id}",
             )(items_embeddings, items_embeddings, items_embeddings, items_mask)
-
```
Comment (rsjeffers) [inline jumanji/training/networks/bin_pack/actor_critic.py]: Add spaces back in so we don't have unnecessary changes wrt the original jumanji

File: jumanji/training/networks/bin_pack/actor_critic.py [hunk 14]
```diff
@@ -187,7 +218,12 @@ def network_fn(observation: Observation) -> chex.Array:
 
         # Outer-product between the embeddings to obtain logits.
         logits = jnp.einsum("...ek,...ik->...ei", ems_embeddings, items_embeddings)
-        logits = jnp.where(observation.action_mask, logits, jnp.finfo(jnp.float32).min)
+        reshaped_action_mask = observation.action_mask.reshape(
```
Comment (rsjeffers) [inline jumanji/training/networks/bin_pack/actor_critic.py]: A small comment explaining what the different axes correspond to would be useful here :)

File: jumanji/training/setup_train.py [hunk 15]
```diff
@@ -88,7 +89,24 @@ def setup_logger(cfg: DictConfig) -> Logger:
...
+            and cfg.env.name not in "upgraded_bin_pack"
+        ):
+            raise ValueError(
+                "Unregistered environment setup not possible for any other argument"
+                f"other than bin_pack, env requested is {cfg.env.name}."
+            )
+        env_settings_dict = {}
+        try:
+            env_settings_dict = {**cfg.env.env_settings}
+        except Exception:
```
Comment (rsjeffers) [inline jumanji/training/setup_train.py]: This try except is very general and might not be catching the right thing. It might be better to check the config for whether you want to apply the constraint or not and do an if based on that

File: jumanji/training/setup_train.py [hunk 16]
```diff
@@ -134,7 +152,7 @@ def _setup_random_policy(  # noqa: CCR001
     cfg: DictConfig, env: Environment
 ) -> RandomPolicy:
     assert cfg.agent == "random"
-    if cfg.env.name == "bin_pack":
+    if cfg.env.name == "bin_pack" or cfg.env.name == "upgraded_bin_pack":
         assert isinstance(env.unwrapped, BinPack)
```
Comment (rsjeffers) [inline jumanji/training/setup_train.py:168]: Shouldn't it be a a constrained binpack class in this case?

File: jumanji/training/setup_train.py [hunk 17]
```diff
@@ -201,7 +219,7 @@ def _setup_actor_critic_neworks(  # noqa: CCR001
     cfg: DictConfig, env: Environment
 ) -> ActorCriticNetworks:
     assert cfg.agent == "a2c"
-    if cfg.env.name == "bin_pack":
+    if cfg.env.name == "bin_pack" or cfg.env.name == "upgraded_bin_pack":
         assert isinstance(env.unwrapped, BinPack)
```
Comment (rsjeffers) [inline jumanji/training/setup_train.py:235]: Same comment as above

File: jumanji/environments/packing/bin_pack/viewer.py [hunk 18]
```diff
@@ -274,3 +274,85 @@ def _get_used_volume(self, state: State) -> float:
             if placed
         )
         return used_volume
+
+
+class ConstrainedBinPackViewer(BinPackViewer):
```
Comment (rsjeffers) [inline jumanji/environments/packing/bin_pack/viewer.py:279]: What's the difference compared to the usual one here?

File: jumanji/environments/packing/bin_pack/generator.py [hunk 19]
```diff
@@ -875,3 +880,74 @@ def body_fn(
...
+        solution.items = Item(
+            x_len=jnp.array([x_len, x_len, y_len, y_len, z_len, z_len]),
+            y_len=jnp.array([y_len, z_len, x_len, z_len, y_len, x_len]),
+            z_len=jnp.array([z_len, y_len, z_len, x_len, x_len, y_len]),
+        )
+        solution.items_mask = jnp.array(
+            [
+                [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
+                [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
+                [0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 0],
```
Comment (rsjeffers) [inline jumanji/environments/packing/bin_pack/generator.py:1283]: Is this because eg. the first item can only be placed in the first 2 orientations?

File: jumanji/environments/packing/bin_pack/env_test.py [hunk 20]
```diff
@@ -38,70 +41,6 @@
 from jumanji.types import TimeStep
 
 
-@pytest.fixture
```
Comment (rsjeffers) [inline jumanji/environments/packing/bin_pack/env_test.py]: Is there a way to add your additions in a way that doesn't show so many changes? (like at the end of the file so it looks like an addition rather than a delete and an addition of the same thing - it would make it easier to review and easier to merge back into jumanji if the team would like the new features

File: jumanji/environments/packing/bin_pack/env.py [hunk 21]
```diff
@@ -701,3 +714,271 @@ def inclusion_check(ems: EMS, ems_mask: chex.Array) -> chex.Array:
...
+
+        # Pack the item if the provided action is valid.
+        next_state = jax.lax.cond(
+            action_is_valid,
+            lambda s: self._pack_item(s, ems_id, item_id, orientation),
+            lambda s: s,
+            state,
+        )
+        # Make the observation.
+        next_state, observation, extras = self._make_observation_and_extras(next_state)
```
Comment (rsjeffers) [inline jumanji/environments/packing/bin_pack/env.py:918]: I think it would be cleaner to make the observation flat before passing it to the network (have a flat spec, and flatten it in make_observation_and_extras.

File: jumanji/environments/packing/bin_pack/env_test.py (hunk 20 above)
Comment (chouakifares) [reply jumanji/environments/packing/bin_pack/env_test.py]: The problem here is that the isn't just an addition but more of a refactor, since the tests were originally not in a class. I could try to restore the wy that file was before and add only my tests into it if you think that would be easier to review and merge.

File: jumanji/environments/packing/bin_pack/reward.py (hunk 8 above)
Comment (chouakifares) [reply jumanji/environments/packing/bin_pack/reward.py]: Done ✅

File: jumanji/environments/packing/bin_pack/generator.py (hunk 19 above)
Comment (chouakifares) [reply jumanji/environments/packing/bin_pack/generator.py:905]: Yes ,  I will add a comment explaining it.

File: jumanji/environments/packing/bin_pack/types.py (hunk 10 above)
Comment (chouakifares) [reply jumanji/environments/packing/bin_pack/types.py:49]: Done ✅

File: jumanji/training/setup_train.py (hunk 17 above)
Comment (chouakifares) [reply jumanji/training/setup_train.py:235]: Done ✅

File: jumanji/environments/packing/bin_pack/types.py (hunk 9 above)
Comment (chouakifares) [reply jumanji/environments/packing/bin_pack/types.py]: I added a comment explaining why we need the extra dimension.

File: jumanji/training/networks/bin_pack/actor_critic.py (hunk 11 above)
Comment (chouakifares) [reply jumanji/training/networks/bin_pack/actor_critic.py]: Removed all the changes I made on the actor-critic file since flattening the observations of the agent is enough to make the training work.

File: jumanji/training/networks/bin_pack/actor_critic.py (hunk 12 above)
Comment (chouakifares) [reply jumanji/training/networks/bin_pack/actor_critic.py]: Yes but I removed this function from the actor-critic file , it is now defined inside the method _make_observations_and_extra.

File: jumanji/training/networks/bin_pack/actor_critic.py (hunk 13 above)
Comment (chouakifares) [reply jumanji/training/networks/bin_pack/actor_critic.py]: I restored the original actor_critic file

File: jumanji/training/networks/bin_pack/actor_critic.py (hunk 14 above)
Comment (chouakifares) [reply jumanji/training/networks/bin_pack/actor_critic.py]: Removed this

File: jumanji/training/setup_train.py (hunk 16 above)
Comment (chouakifares) [reply jumanji/training/setup_train.py:168]: done ✅

File: jumanji/environments/packing/bin_pack/env.py (hunk 21 above)
Comment (chouakifares) [reply jumanji/environments/packing/bin_pack/env.py:918]: Done ✅

File: jumanji/environments/packing/bin_pack/viewer.py (hunk 18 above)
Comment (chouakifares) [reply jumanji/environments/packing/bin_pack/viewer.py:279]: The methods `create_entities` and `add_overlay` and `_get_used_volume` have to be changed

File: jumanji/training/setup_train.py (hunk 15 above)
Comment (chouakifares) [reply jumanji/training/setup_train.py]: I fixed this.

File: jumanji/environments/packing/bin_pack/generator.py (hunk 19 above)
Comment (chouakifares) [reply jumanji/environments/packing/bin_pack/generator.py:1283]: Yes, I added a comment explaining this.

File: jumanji/environments/packing/bin_pack/env_test.py (hunk 20 above)
Comment (rsjeffers) [reply jumanji/environments/packing/bin_pack/env_test.py]: Yes I don't think we should be changing other people's jumanji code if possible

Comment (rsjeffers) [reply jumanji/environments/packing/bin_pack/env_test.py]: You can save it locally and create an issue on the main jumanji if you think it is something that should be refactored and it can be done in its own issue :)

File: jumanji/environments/packing/bin_pack/viewer.py (hunk 18 above)
Comment (rsjeffers) [reply jumanji/environments/packing/bin_pack/viewer.py:279]: Could you add a docstring explaining this?

File: jumanji/environments/packing/bin_pack/types.py (hunk 9 above)
Comment (rsjeffers) [reply jumanji/environments/packing/bin_pack/types.py]: Why is items_placed (6,max_num_items,6) and not (6,max_num_items,)

Comment (chouakifares) [reply jumanji/environments/packing/bin_pack/types.py]: It was a typo , I fixed it.

File: jumanji/environments/packing/bin_pack/viewer.py (hunk 18 above)
Comment (chouakifares) [reply jumanji/environments/packing/bin_pack/viewer.py:279]: I added a docstring at the beginning of the class definition.

File: jumanji/environments/packing/bin_pack/types.py (hunk 9 above)
Comment (chouakifares) [reply jumanji/environments/packing/bin_pack/types.py]: Restored the file to its original state then added my own tests , should be easier to review now.

File: jumanji/environments/packing/bin_pack/env_test.py (hunk 20 above)
Comment (rsjeffers) [reply jumanji/environments/packing/bin_pack/env_test.py]: There still seems to be a lot of code removed - have you pushed your latest version here?

File: jumanji/environments/packing/bin_pack/viewer.py (hunk 18 above)
Comment (rsjeffers) [reply jumanji/environments/packing/bin_pack/viewer.py:279]: I can't see the docstring is the latest version pushed?

File: jumanji/training/networks/bin_pack/actor_critic.py (hunk 11 above)
Comment (rsjeffers) [reply jumanji/training/networks/bin_pack/actor_critic.py]: It still seems to be there?

File: configs/env/bin_pack.yaml [hunk 22]
```diff
@@ -26,7 +26,6 @@ training:
 evaluation:
     eval_total_batch_size: 2
     greedy_eval_total_batch_size: 2
-
```
Comment (rsjeffers) [inline configs/env/bin_pack.yaml]: could you undo this change please?

File: jumanji/environments/packing/bin_pack/env_test.py [hunk 23]
```diff
@@ -450,7 +450,6 @@ def test_bin_pack__optimal_policy_toy_instance(
             # assert not timestep.extras["invalid_action"]
             assert not jnp.any(timestep.extras["invalid_ems_from_env"])
         if timestep.extras is not None:
-            assert timestep.extras["volume_utilization"] == 1
             assert timestep.extras["ratio_packed_items"] == 1
```
Comment (rsjeffers) [inline jumanji/environments/packing/bin_pack/env_test.py:454]: Why is this no longer 1? This should still pass for the toy instance right?

Comment (chouakifares) [reply jumanji/environments/packing/bin_pack/env_test.py:453]: No it's not supposed to pass, it was commented in the previous commits (and I was supposed to delete that line), I don't know in which commit I accidentally uncommented it. It's not supposed to pass for the toy instance because we have only 3 relatively small items which aren't enough to fill the bin.

File: jumanji/environments/packing/bin_pack/reward.py [hunk 24]
```diff
@@ -112,7 +121,13 @@ def __call__(
...
         state.items = cast(ValuedItem, state.items)
-        _, item_id = action
+        if len(action) == 2:
+            _, item_id = action
+            chosen_item_value = tree_slice(state.items, item_id).value
+        elif len(action) == 3:
+            orientation, _, item_id = action
+            chosen_item_value = tree_slice(state.items, (orientation, item_id)).value
+
         chosen_item_value = tree_slice(state.items, item_id).value
```
Comment (rsjeffers) [inline jumanji/environments/packing/bin_pack/reward.py]: I think this line needs to be removed? Are there not tests that fail due to this?

Comment (chouakifares) [reply jumanji/environments/packing/bin_pack/reward.py]: You mean line 131 ? I removed it.

File: jumanji/environments/packing/bin_pack/env_test.py (hunk 23 above)
Comment (chouakifares) [reply jumanji/environments/packing/bin_pack/env_test.py:454]: I had a bug with the action mask , I fixed it

File: configs/env/bin_pack.yaml (hunk 22 above)
Comment (chouakifares) [reply configs/env/bin_pack.yaml]: done ✅

**Review by rsjeffers — APPROVED** [review]
Ok to merge! thanks
